test/metabase/query_processor/middleware/resolve_joined_fields_test.clj

```diff
@@ -214,3 +214,45 @@
                  :aggregation  [[:count]]
                  :breakout     [$products.id]
                  :limit        5})))))))
+
+(deftest do-not-rewrite-top-level-clauses-if-field-is-from-source-table-or-query
+  (testing (str "Do not add `:join-alias` to top-level `:field` clauses if the Field could come from the "
+                "`:source-table` or `:source-query` (#18502)")
+    (mt/dataset sample-dataset
+      (is (query= (mt/mbql-query people
+                    {:source-query {:source-table $$people
+                                    :breakout     [!month.created_at]
+                                    :aggregation  [[:count]]
+                                    :order-by     [[:asc !month.created_at]]}
+                     :joins        [{:source-query {:source-table $$people
+                                                    :breakout     [!month.birth_date]
+                                                    :aggregation  [[:count]]
+                                                    :order-by     [[:asc !month.birth_date]]}
+                                     :alias        "Q2"
+                                     :condition    [:= !month.created_at !month.&Q2.birth_date]
+                                     :fields       [&Q2.birth_date &Q2.*count/BigInteger]
+                                     :strategy     :left-join}]
+                     :fields       [!default.created_at
+                                    *count/BigInteger
+                                    &Q2.birth_date
+                                    &Q2.*count/BigInteger]
+                     :limit        3})
+                  (wrap-joined-fields
+                   (mt/mbql-query people
+                     {:source-query {:source-table $$people
+                                     :breakout     [!month.created_at]
+                                     :aggregation  [[:count]]
+                                     :order-by     [[:asc !month.created_at]]}
+                      :joins        [{:source-query {:source-table $$people
+                                                     :breakout     [!month.birth_date]
+                                                     :aggregation  [[:count]]
+                                                     :order-by     [[:asc !month.birth_date]]}
+                                      :alias        "Q2"
+                                      :condition    [:= !month.created_at !month.&Q2.birth_date]
+                                      :fields       [&Q2.birth_date &Q2.*count/BigInteger]
+                                      :strategy     :left-join}]
+                      :fields       [!default.created_at
+                                     *count/BigInteger
+                                     &Q2.birth_date
+                                     &Q2.*count/BigInteger]
+                      :limit        3})))))))
```

Comment:
Are these two MBQL forms identical (except the latter is wrapped within the `wrap-joined-fields` call?)?  Is there any value in deduplicating them?

Reply:
```clj
*count/BigInteger
=>
[:field "count" {:base-type :type/BigInteger}]

&Q2.*count/BigInteger
[:field "count" {:base-type :type/BigInteger, :join-alias "Q2"}]
```

they are two different fields, the first one is from the source query while the latter is from the join `Q2`

Reply:
If we don't sort on both then some databases return results in arbitrary order which is not good for having my test pass